### PR TITLE
Explicitly retry failed docker pulls

### LINKED ISSUE #63869: Make docker pull more robust
The Elasticsearch `Dockerfile` tries to make network operations more robust in case of transient problems by retrying e.g. `wget` or `yum` in a loop. However, the ES build process doesn't have any such retry mechanism around pulling Docker images in the first place. We need to rework the build to explicitly pull the requires images, and add a retry loop in case of failure.

### PR DESCRIPTION
**Pull request by pugnascotia**
Closes #63869. Perform `docker pull` explicitly instead of as part of
`docker build`, and wrap it in a retry loop. This is an attempt to make
the build more resilient to transient errors.

### CODE REVIEW

File: buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerBuildTask.java [hunk 1]
```diff
@@ -111,26 +127,57 @@ public DockerBuildAction(ExecOperations execOperations) {
...
+                        spec.executable("docker");
+                        spec.args("pull");
+                        spec.args(baseImage);
+                    });
+
+                    break;
+                } catch (Exception e) {
+                    LOGGER.warn("Attempt {}/{} to pull Docker base image {} failed", attempt, maxAttempts, baseImage);
+
+                    if (attempt == 5) {
```
Comment (rjernst) [inline buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerBuildTask.java:150]: This should probably compare to `maxAttempts`?

**Review by rjernst — APPROVED** [review]
LGTM

**Review by mark-vieira — APPROVED** [review]
:+1:

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/eql-correctness

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by pugnascotia** [comment]
@elasticmachine update branch

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-unix due to timeout

**Comment by pugnascotia** [comment]
Backported to `7.x` in ad2637c229b

**Pull request merged**